### PR TITLE
LDAP: Optional user name attribute specification

### PR DESCRIPTION
**Pull request by nanoant**
Consider following LDAP search query example:

    (&(objectClass=Person)(|(uid=%s)(mail=%s)))

Right now on first login attempt Gogs will use the text supplied on login form
as the newly created user name. In example query above the text matches against
both e-mail or user name. So if user puts the e-mail then the new Gogs user
name will be e-mail which may be undesired.

Using optional user name attribute setting we can explicitly say we want Gogs
user name to be certain LDAP attribute eg. `uid`, so even user will use e-mail
to login 1st time, the new account will receive correct user name.

### CODE REVIEW

File: models/login.go [hunk 1]
```diff
@@ -242,17 +242,30 @@ func LoginUserLDAPSource(u *User, name, passwd string, source *LoginSource, auto
 	}
 
 	// Fallback.
+	if len(name) == 0 {
+		name = loginName
+	}
 	if len(mail) == 0 {
 		mail = fmt.Sprintf("%s@localhost", name)
 	}
+	var fullName string
```
Comment (Unknwon) [inline models/login.go]: I think it would be better to put into a function like `composeFullName(fn,sn) string`

Comment (nanoant) [inline models/login.go]: @Unknwon agreed, I like you approach of having clean reusable code! Fixing that.

File: models/login.go [hunk 2]
```diff
@@ -260,6 +263,18 @@ func LoginUserLDAPSource(u *User, name, passwd string, source *LoginSource, auto
 	return u, CreateUser(u)
 }
 
+func composeFullName(firstName, surename, userName string) string {
```
Comment (Unknwon) [inline models/login.go]: As you return in every case branch, you do not need `else if` anymore.

**Comment by Unknwon** [comment]
LGTM.

**Comment by wqweto** [comment]
@nanoant FYI, search query has to use fmt's `explicit argument indexes`  to make it work
```
(&(objectClass=Person)(|(uid=%[1]s)(mail=%[1]s)))
```

**Comment by nanoant** [comment]
@wqweto Oh.. thanks, completely forgot about that one.

**Pull request merged**